crates/chain-worker/src/service.rs

```diff
@@ -0,0 +1,296 @@
+
+        let cur_tip = match init_state.get_declared_final_epoch().cloned() {
+            Some(epoch) => epoch.to_block_commitment(),
+            None => L2BlockCommitment::new(
+                0,
+                *init_state.sync().expect("after genesis").genesis_blkid(),
+            ),
+        };
+
+        Ok(cur_tip)
+    }
+
+    fn initialize_with_tip(&mut self, cur_tip: L2BlockCommitment) -> anyhow::Result<()> {
+        let blkid = *cur_tip.blkid();
+        info!(%blkid, "initializing chain worker");
+
+        self.cur_tip = cur_tip;
+        self.initialized = true;
+
+        Ok(())
+    }
+
+    /// Prepares context for a block we're about to execute.
+    fn prepare_block_context<'w>(
+        &'w self,
+        _l2bc: &L2BlockCommitment,
+    ) -> WorkerResult<WorkerExecCtxImpl<'w, W>> {
+        Ok(WorkerExecCtxImpl {
+            worker_context: &self.context,
+        })
+    }
+
+    fn try_exec_block(&mut self, block: &L2BlockCommitment) -> WorkerResult<()> {
+        self.check_initialized()?;
+
+        let context = &self.context;
+        let chain_exec = &self.chain_exec;
+
+        let blkid = block.blkid();
+        debug!(%blkid, "Trying to execute block");
+
+        let bundle = context
+            .fetch_block(block.blkid())?
+            .ok_or(WorkerError::MissingL2Block(*block.blkid()))?;
+
+        let is_epoch_terminal = !bundle.body().l1_segment().new_manifests().is_empty();
+
+        let parent_blkid = bundle.header().header().parent();
+        let parent_header = context
+            .fetch_header(parent_blkid)?
+            .ok_or(WorkerError::MissingL2Block(*parent_blkid))?;
+
+        self.exec_ctl_handle
+            .try_exec_el_payload_blocking(*block)
+            .map_err(|_| WorkerError::InvalidExecPayload(*block))?;
+
+        let header_ctx = strata_chaintsn::context::L2HeaderAndParent::new(
+            bundle.header().header().clone(),
+            *parent_blkid,
+            parent_header,
+        );
+
+        let exec_ctx = self.prepare_block_context(block)?;
+
+        let output = chain_exec.verify_block(&header_ctx, bundle.body(), &exec_ctx)?;
+
+        if is_epoch_terminal {
+            debug!(%is_epoch_terminal);
+            self.handle_complete_epoch(block.blkid(), bundle.block(), &output)?;
+        }
+
+        self.context.store_block_output(block.blkid(), &output)?;
+
+        Ok(())
+    }
+
+    /// Takes the block and post-state and inserts database entries to reflect
+    /// the epoch being finished on-chain.
+    ///
+    /// There's some bookkeeping here that's slightly weird since in the way it
+    /// works now, the last block of an epoch brings the post-state to the new
+    /// epoch.  So the epoch's final state actually has cur_epoch be the *next*
+    /// epoch.  And the index we assign to the summary here actually uses the
+    /// "prev epoch", since that's what the epoch in question is here.
```

Comment:
```suggestion
    /// There's some bookkeeping here that's slightly weird since in the way it
    /// works now, the last block of an epoch brings the post-state to the new
    /// epoch. So the epoch's final state actually has cur_epoch be the *next*
    /// epoch. And the index we assign to the summary here actually uses the
    /// "prev epoch", since that's what the epoch in question is here.
```